libraries/AP_OpenDroneID/AP_OpenDroneID.cpp

```diff
@@ -0,0 +1,611 @@
+        MAV_ODID_VER_ACC mavoutput;     // mavlink enum output
+    } vertical_accuracy_table[] = {
+        { 1.0,  MAV_ODID_VER_ACC_1_METER},
+        { 3.0,  MAV_ODID_VER_ACC_3_METER},
+        {10.0,  MAV_ODID_VER_ACC_10_METER},
+        {25.0,  MAV_ODID_VER_ACC_25_METER},
+        {45.0,  MAV_ODID_VER_ACC_45_METER},
+        {150.0, MAV_ODID_VER_ACC_150_METER},
+    };
+
+    for (auto elem : vertical_accuracy_table) {
+        if (accuracy < elem.accuracy) {
+            return elem.mavoutput;
+        }
+    }
+
+    // Should not reach this
+    return MAV_ODID_VER_ACC_UNKNOWN;
+}
+
+/**
+* This converts a speed accuracy float value to the corresponding enum
+*
+* @param Accuracy The speed accuracy in m/s
+* @return Enum value representing the accuracy
+*/
+MAV_ODID_SPEED_ACC AP_OpenDroneID::create_enum_speed_accuracy(float accuracy) const
+{
+    // Out of bounds return UKNOWN flag
+    if (accuracy < 0.0 || accuracy >= 10.0) {
+        return MAV_ODID_SPEED_ACC_UNKNOWN;
+    }
+
+    if (accuracy < 0.3) {
+        return MAV_ODID_SPEED_ACC_0_3_METERS_PER_SECOND;
+    } else if (accuracy < 1.0) {
+        return MAV_ODID_SPEED_ACC_1_METERS_PER_SECOND;
+    } else if (accuracy < 3.0) {
+        return MAV_ODID_SPEED_ACC_3_METERS_PER_SECOND;
+    } else if (accuracy < 10.0) {
+        return MAV_ODID_SPEED_ACC_10_METERS_PER_SECOND;
+    }
+
+    // Should not reach this
+    return MAV_ODID_SPEED_ACC_UNKNOWN;
+}
+
+/**
+* This converts a timestamp accuracy float value to the corresponding enum
+*
+* @param Accuracy The timestamp accuracy in seconds
+* @return Enum value representing the accuracy
+*/
+MAV_ODID_TIME_ACC AP_OpenDroneID::create_enum_timestamp_accuracy(float accuracy) const
+{
+    // Out of bounds return UKNOWN flag
+    if (accuracy < 0.0 || accuracy >= 1.5) {
+        return MAV_ODID_TIME_ACC_UNKNOWN;
+    }
+
+    static const MAV_ODID_TIME_ACC mavoutput [15] = {
+        MAV_ODID_TIME_ACC_0_1_SECOND,
+        MAV_ODID_TIME_ACC_0_2_SECOND,
+        MAV_ODID_TIME_ACC_0_3_SECOND,
+        MAV_ODID_TIME_ACC_0_4_SECOND,
+        MAV_ODID_TIME_ACC_0_5_SECOND,
+        MAV_ODID_TIME_ACC_0_6_SECOND,
+        MAV_ODID_TIME_ACC_0_7_SECOND,
+        MAV_ODID_TIME_ACC_0_8_SECOND,
+        MAV_ODID_TIME_ACC_0_9_SECOND,
+        MAV_ODID_TIME_ACC_1_0_SECOND,
+        MAV_ODID_TIME_ACC_1_1_SECOND,
+        MAV_ODID_TIME_ACC_1_2_SECOND,
+        MAV_ODID_TIME_ACC_1_3_SECOND,
+        MAV_ODID_TIME_ACC_1_4_SECOND,
+        MAV_ODID_TIME_ACC_1_5_SECOND,
+    };
+
+    for (int8_t i = 1; i <= 15; i++) {
+        if (accuracy <= 0.1 * i) {
+            return mavoutput[i-1];
+        }
+    }
+
+    // Should not reach this
+    return MAV_ODID_TIME_ACC_UNKNOWN;
+}
+
+// make sure value is within limits of remote ID standard
+uint16_t AP_OpenDroneID::create_speed_horizontal(uint16_t speed) const
+{
+    if (speed > ODID_MAX_SPEED_H) { // constraint function can't be used, because out of range value is invalid
```

Comment:
This seems like a very odd construction, ODID_MAX_SPEED_H is a float, that we compare against a `uint16_t` argument that is truncating our floating point representation. I suspect that the `speed` argument is supposed to be a float, and the return also is based on the defines, and the fact that anything else truncates to meter resolution.

Why not scale in here and have this just return the correctly output scalled cm/s uint16_t?

This looks like a common problem on these helpers.